mgmt/osdp: Replace __ASSERT() with an explicit if

sidcha opened the pull request:
Commit c7fec71193a ("mgmt/osdp: Add length checks for commands and replies") attempted to remove code duplication by adding a macro to perform a length check. At the time, a CI linter did not like macros with control flow so the code was switched to a method which called __ASSERT() on this condition.

The __ASSERT() macro is a nop if CONFIG_ASSERT=n (which is the default) and causes the buffer access to be unguarded which may lead to OOB accesses. This patch fixes the issue by reintroducing the if check.

Fixes: c7fec71193a19f6be1a2adca8cf7753cd7103c78.

sidcha committed:
mgmt/osdp: Replace __ASSERT() with an explicit if

Commit c7fec71193a ("mgmt/osdp: Add length checks for commands and
replies") attempted to remove code duplication by adding a macro to perform
a length check. At the time, a CI linter did not like macros with control
flow so the code was switched to a method which called __ASSERT() on this
condition.

The __ASSERT() macro is a nop if CONFIG_ASSERT=n (which is the default) and
causes the buffer access to be unguarded which may lead to OOB accesses.
This patch fixes the issue by reintroducing the if check.

Fixes: c7fec71193a19f6be1a2adca8cf7753cd7103c78.
Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha commented:
cc: @ceolin

r2r0 commented on subsys/mgmt/osdp/src/osdp_cp.c:
```diff
@@ -114,10 +114,13 @@ int osdp_extract_address(int *address)
 {
-	__ASSERT(need < have, "OOM at build command: need:%d have:%d",
-		 need, have);
+	if (need >= have) {
```
Why `need > have` is not sufficient (negative) check condition?
If there is some assumption for usage of the last byte buffer then a small comment will be very appreciated here (the same for PD part).

sidcha replied:
`need > have` would be fine here. I was retaining the existing condition to have no functional change in this commit. The condition in the `__ASSERT()` comes from a time when the command/reply ID (1 byte) was not part of `need`.

sidcha committed:
mgmt/osdp: Fix off-by-one in buf len checks

Initially, the command/reply ID byte was not part of the data length
macros. But later, when it was changed to include it, the buffer length
checks was not adjusted. Due to this, we were not using the last byte in
the buffer. Fix this issue by correcting the condition.

Signed-off-by: Siddharth Chandrasekaran <sidcha.dev@gmail.com>

sidcha commented:
@ceolin Added another commit to address the issue pointed out by @r2r0. Please take another look.

sidcha commented:
@ceolin Can you please take another look?

r2r0 commented:
@ceolin Would you mind to taking a quick look?

ceolin commented on subsys/mgmt/osdp/src/osdp_cp.c:
```diff
@@ -137,42 +140,60 @@ static int cp_build_command(struct osdp_pd *pd, uint8_t *buf, int max_len)
 	switch (pd->cmd_id) {
 	case CMD_POLL:
-		assert_buf_len(CMD_POLL_LEN, max_len);
+		if (!check_buf_len(CMD_POLL_LEN, max_len)) {
```
Is this an unrecoverable error ? If not you should use `LOG_WRN` imho.
Also, `check_buf_len` is essentially just a branch. Wouldn't be better do the check here, and if it is false use a goto and log it with more information ? Something like:
```
case CMD_POLL:
        if (CMD_POLL_LEN > max_len) {
                goto err;
        }
.....


error:
	LOG_WRN("OOM at build command (%d) fails\n", pd->cmd_id);
	return OSDP_CP_ERR_GENERIC;
```

sidcha replied:
This error should technically be unreachable since the both peers agree on TX/RX buffer size. If we _do_ hit, we should treat it as unrecoverable error (PD goes offline).

> Wouldn't be better do the check here, and if it is false [..]

When the method is `static inline`-ed. An optimising compiler should produce the same effect IIRC. Also, logging `need` and `have` is more important (than the command/reply ID as they are logged earlier).

ceolin reviewed (approved):
I made a suggestion, but the fix is correct imho.
